internal/pkg/cli/deploy/workload.go

```diff
@@ -367,24 +392,86 @@ func (d *workloadDeployer) uploadContainerImages(out *UploadArtifactsOutput) err
 	if err != nil {
 		return fmt.Errorf("login to image repository: %w", err)
 	}
+
+	var mu sync.Mutex
 	out.ImageDigests = make(map[string]ContainerImageIdentifier, len(buildArgsPerContainer))
+
+	// counter for indexing the labeled output buffers.
+	bufferIdx := 0
+	// An array of buffers, one for each container image build and push output.
+	labeledBuffers := make([]*syncbuffer.LabeledSyncBuffer, len(buildArgsPerContainer))
+
+	// create a context and an error group.
+	g, ctx := errgroup.WithContext(context.Background())
+
+	// create a new cursor and hide it.
+	cursor := cursor.New()
+	cursor.Hide()
+
 	for name, buildArgs := range buildArgsPerContainer {
+		// create a copy of loop variables to avoid data race.
+		name := name
+		buildArgs := buildArgs
+
+		// create a pipe for streaming the build and push output.
+		pr, pw := io.Pipe()
+
 		buildArgs.URI = uri
+
+		// generate build args slice for docker build label.
 		buildArgsList, err := buildArgs.GenerateDockerBuildArgs(dockerengine.New(exec.NewCmd()))
 		if err != nil {
-			return fmt.Errorf("generate docker build args: %w", err)
-		}
-		// TODO(adi) handle this log in syncBuffer's Print function
-		log.Infof("Building your container image: docker %s\n", strings.Join(buildArgsList, " "))
-		digest, err := d.repository.BuildAndPush(buildArgs)
-		if err != nil {
-			return fmt.Errorf("build and push image: %w", err)
+			return fmt.Errorf("generate docker build args of image %s: %w", name, err)
 		}
-		out.ImageDigests[name] = ContainerImageIdentifier{
-			Digest:            digest,
-			CustomTag:         d.image.CustomTag,
-			GitShortCommitTag: d.image.GitShortCommitTag,
+		buf := syncbuffer.New()
+		labeledBuffer := buf.WithLabel(fmt.Sprintf("Building your container image %s: docker %s", name, strings.Join(buildArgsList, " ")))
+		labeledBuffers[bufferIdx] = labeledBuffer
+		bufferIdx++
+
+		g.Go(func() error {
+			defer pw.Close()
+			digest, err := d.repository.BuildAndPush(ctx, buildArgs, pw)
+			if err != nil {
+				return fmt.Errorf("build and push image %s: %w", name, err)
+			}
+			mu.Lock()
+			out.ImageDigests[name] = ContainerImageIdentifier{
+				Digest:            digest,
+				CustomTag:         d.image.CustomTag,
+				GitShortCommitTag: d.image.GitShortCommitTag,
+			}
+			mu.Unlock()
+			return nil
+		})
+		g.Go(func() error {
+			if err := buf.Copy(pr); err != nil {
```

Comment:
yeah that's true. When `buf.Copy ` is called, it will block until `pw.Close()` is invoked, at which point it will return the final error (if any) encountered during the copy. So `pw.Close()` is used to signal the end of the build process and allow the copy process to finish.

Reply:
gotcha! ok so the internal bytes.Buffer in the syncbuffer does get populated while `Copy` is still running that's why when we call `buf.lines()` it's able to return to us partial results.

This makes me think:
- Is there a data race when `buf.Copy` is happening (because of `io.Copy` that's writing to the bytes.Buffer) in one Go routine and from another Go routine we are calling `lines()` to read from the same underlying bytes.Buffer?

It might be interesting to chat w @dannyrandall on it

Reply:
I think we want something like this to be concurrency-safe:
https://github.com/aws/copilot-cli/blob/0ea2c4ba4b01dd72c74117803e5f9ea7c7c930d2/internal/pkg/term/syncbuffer/syncbuffer.go#L22-L27

```go
 // New creates and returns a new SyncBuffer object with an initialized 'done' channel. 
 func New() *SyncBuffer { 
        pr, pw := io.Pipe()
 	return &SyncBuffer{ 
 		done: make(chan struct{}), 
                r: pr,
                w: pw
 	} 
 } 
 
func (buf *SyncBuffer) Copy(r io.Reader) error {
	defer buf.MarkDone()
	_, err := io.Copy(&buf.w, r)
	if err != nil && !errors.Is(err, io.EOF) {
		return err
	}
	return nil
}

func (buf *SyncBuffer) lines() []string {
	lines := strings.NewReader(buf.r).Strings()
	if len(lines) == 0 {
		return nil
	}
	return splitLinesAndTrimSpaces(lines)
}
```

Reply:
I already made that change  over here.
https://github.com/aws/copilot-cli/pull/4760/files#diff-b12c92294403172e636dd4c441963a6ff7d684f316c919f6a491f76d37a65a37R68

If I am directly writing to  internal bytes.Buffer something like this  `_, err := io.Copy(&buf.buf, r)`. I got warning of Datarace. So I changed it to `_, err := io.Copy(buf, r)`

Reply:
Very nice!